### PR TITLE
Define StatusOr, Status, and StatusCode datatypes

### LINKED ISSUE #33: Define StatusOr, Status, and StatusCode datatypes for passing error codes
Split from #9.

### PR DESCRIPTION
**Pull request by ZacharyEspiritu**
Resolves #33.

Defines `StatusCode`, `Status`, and `StatusOr<T>` definitions. Also, defines the other half of the codebase as `namespace backing` (but other naming suggestions welcome).

- Is there something that needs to be done to the licensing headers of the files to signal that modifications were made to them?

### CODE REVIEW

File: src/backing/status/status.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  bool ok() const {
+    return error_code_ == StatusCode::kOk;
+  }
+  StatusCode error_code() const {
+    return error_code_;
+  }
+  StatusCode code() const {
+    return error_code_;
+  }
+  std::string error_message() const {
```
Comment (ZacharyEspiritu) [inline src/backing/status/status.h:71]: This could be a `std::string_view`, maybe?

**Comment by ajfeldma-google** [comment]
> Resolves #33.
> 
> Defines `StatusCode`, `Status`, and `StatusOr<T>` definitions. Also, defines the other half of the codebase as `namespace backing` (but other naming suggestions welcome).
> 
> * Is there something that needs to be done to the licensing headers of the files to signal that modifications were made to them?

I think so. The [Apache license](https://www.apache.org/licenses/LICENSE-2.0) (the one that the Tensorflow code uses) requires it. Maybe the thing to do is to add a new header at the top but then include the previous header below it along with a list of modifications.

**Comment by ajfeldma-google** [comment]
Probably not. Generally, returning a string_view isn't a good idea because
the lifetime of the string that backs the string_view isn't easy to keep
track of.

File: src/backing/status/status_or.h [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+#include <new>
+#include <string>
+#include <utility>
+
+#include "src/backing/status/status.h"
+#include "src/backing/status/status_code.h"
+
+namespace kmsengine {
+namespace backing {
```
Comment (tonyliaoss) [inline src/backing/status/status_or.h:117]: Where does the name "backing" come from?

I wouldn't be opposed to just leaving the statuses in the common namespace.

Comment (ZacharyEspiritu) [reply src/backing/status/status_or.h:117]: It's the name for the `libkms` part of the codebase. Given the previous discussion on not including "lib" in the namespace names, I figured `namespace kmsengine { namespace kms { ... } }` might be confusing and tried to pick something else.

I think leaving in the common namespace is a good idea (as I've found while trying to use these types) - I've just been trying to make a clear distinction between which parts of the codebase are only allowed to depend on OpenSSL and which parts of the codebase are allowed to depend on other dependencies (that aren't OpenSSL).

**Comment by ZacharyEspiritu** [comment]
Closing this since we'll defer definitions of `Status`, etc. to the `@com_google_google_cloud_cpp//google/cloud:google_cloud_cpp_common` rule from [googleapis/google-cloud-cpp](https://github.com/googleapis/google-cloud-cpp).

**Pull request closed without merging**